sw/device/silicon_creator/lib/base/sec_mmio_unittest.cc

```diff
@@ -0,0 +1,162 @@
+using ::testing::Eq;
+using ::testing::Test;
+
+class SecMmioTest : public mask_rom_test::MaskRomTest {
+ protected:
+  void SetUp() override {
+    *shutdown_count_ = 0;
+    sec_mmio_init(+[] { ++__shutdown_count; });
+  }
+
+  uint32_t GetShutDownCount() { return *shutdown_count_; }
+
+  sec_mmio_ctx_t *ctx_ = &::sec_mmio_ctx;
+  uint32_t *shutdown_count_ = &::__shutdown_count;
+  mask_rom_test::MockAbsMmio mmio_;
+};
+
+TEST_F(SecMmioTest, Initialize) {
+  EXPECT_EQ(ctx_->check_count, 0);
+  EXPECT_EQ(ctx_->expected_write_count, 0);
+  EXPECT_EQ(ctx_->last_index, 0);
+  EXPECT_EQ(ctx_->write_count, 0);
+  EXPECT_THAT(ctx_->addrs, Each(Eq(UINT32_MAX)));
+}
+
+TEST_F(SecMmioTest, Read32OrDie) {
+  EXPECT_ABS_READ32(mmio_, 0, 0x12345678);
+  EXPECT_ABS_READ32(mmio_, 0, 0x12345678);
+  EXPECT_EQ(sec_mmio_read32(0), 0x12345678);
+
+  EXPECT_ABS_READ32(mmio_, 4, 0x87654321);
+  EXPECT_ABS_READ32(mmio_, 4, 0x87654321);
+  EXPECT_EQ(sec_mmio_read32(4), 0x87654321);
+
+  EXPECT_ABS_READ32(mmio_, 0, 0x87654321);
+  EXPECT_ABS_READ32(mmio_, 0, 0x87654321);
+  EXPECT_EQ(sec_mmio_read32(0), 0x87654321);
+
+  // Two of the operations were targeting the same offset, so we only expect two
+  // operations and zero shutdown attempts.
+  EXPECT_EQ(ctx_->write_count, 0);
+  EXPECT_EQ(GetShutDownCount(), 0);
+  EXPECT_EQ(ctx_->last_index, 2);
+}
+
+TEST_F(SecMmioTest, Read32ORDieSimulatedFault) {
+  EXPECT_ABS_READ32(mmio_, 0, 0x12345678);
+  EXPECT_ABS_READ32(mmio_, 0, 0);
+  EXPECT_EQ(sec_mmio_read32(0), 0x12345678);
+  EXPECT_EQ(ctx_->write_count, 0);
+  EXPECT_EQ(GetShutDownCount(), 1);
+}
+
+TEST_F(SecMmioTest, Write32) {
+  EXPECT_ABS_WRITE32(mmio_, 0, 0x12345678);
+  EXPECT_ABS_READ32(mmio_, 0, 0x12345678);
+  sec_mmio_write32(0, 0x12345678);
+  EXPECT_EQ(ctx_->write_count, 1);
+
+  EXPECT_ABS_WRITE32(mmio_, 4, 0x87654321);
+  EXPECT_ABS_READ32(mmio_, 4, 0x87654321);
+  sec_mmio_write32(4, 0x87654321);
+  EXPECT_EQ(ctx_->write_count, 2);
+
+  EXPECT_ABS_WRITE32(mmio_, 0, 0x87654321);
+  EXPECT_ABS_READ32(mmio_, 0, 0x87654321);
+  sec_mmio_write32(0, 0x87654321);
+  EXPECT_EQ(ctx_->write_count, 3);
+
+  // Two of the operations were targeting the same offset, so we only expect two
+  // operations and zero shutdown attempts.
+  EXPECT_EQ(GetShutDownCount(), 0);
+  EXPECT_EQ(ctx_->last_index, 2);
+}
+
+TEST_F(SecMmioTest, Write32SimulatedFault) {
+  EXPECT_ABS_WRITE32(mmio_, 0, 0x12345678);
+  EXPECT_ABS_READ32(mmio_, 0, 0);
+  sec_mmio_write32(0, 0x12345678);
+  EXPECT_EQ(GetShutDownCount(), 1);
+}
+
+TEST_F(SecMmioTest, CounterInc) {
+  sec_mmio_write_increment(5);
+  EXPECT_EQ(ctx_->expected_write_count, 5);
+
+  sec_mmio_write_increment(10);
+  EXPECT_EQ(ctx_->expected_write_count, 15);
+}
+
+TEST_F(SecMmioTest, CheckValues) {
+  EXPECT_ABS_WRITE32(mmio_, 0, 0x12345678);
+  EXPECT_ABS_READ32(mmio_, 0, 0x12345678);
+  sec_mmio_write32(0, 0x12345678);
+
+  EXPECT_ABS_READ32(mmio_, 0, 0x12345678);
+  sec_mmio_check_values(/*rnd_offset=*/0);
```

Comment:
Are there any tests which use a different rnd_offset?

Reply:
Perhaps a test that hammers it with a range of random values is warranted?

Reply:
Done. Added additional test cases to test different permutations.